### PR TITLE
Port the CalculiX adapter to preCICE v3

### PR DESCRIPTION
**Pull request by MakisH**
See the [porting guide](https://precice.org/couple-your-code-porting-v2-3.html).

- [x] Port `SolverInterface` headers and constructor
- [x] Port `setMeshVertices` -> should need no changes
- [x] Port connectivity (set mesh edges / triangles)
- [x] Merge `Initialize` and `InitializeData`
- [x] Replace action to write initial data with `requiresInitialData`
- [x] Remove `isReadDataAvailable` and `isWriteDataRequired`
- [x] Port checkpointing (requires/fulfilled -> requires)
- [x] Port all methods that use IDs to use names (read/write/...)
- [x] Merge all read/write methods
- [x] Add dt to all methods that read data
- [x] Remove dt return value from initialize/advance

### CODE REVIEW

**Review by BenjaminRodenberg — COMMENTED** [review]
I added a few comments on the time step size etc. below. I'm currently wondering a bit what happens, if CalculiX time step size (i.e. what's provided to the [`DYNAMIC`](https://web.mit.edu/calculix_v2.7/CalculiX/ccx_2.7/doc/ccx/node192.html) keyword) differs from the time window size defined in the preCICE configuration file.

**Comment by fsimonis** [comment]
@BenjaminRodenberg I rewrote the time step computation to make it generally simpler to read.

Instead of normalizing the precice_dt, I now compute the calculix_dt = dtheta * tper, then find the synchronized dt and normalize that again.

This makes the code and the logs easier to understand from a preCICE perspective: 

```
Adjusting time step for transient step
precice_dt = 0.010000, ccx_dt = 0.010000 (dtheta = 0.020000, tper = 0.500000) -> dt = 0.010000 (dtheta = 0.020000)
```

File: adapter/PreciceInterface.c [hunk 1]
```diff
@@ -204,89 +191,87 @@ void Precice_ReadCouplingData(SimulationData *sim)
...
+    for (j = 0; j < interfaces[i]->numReadData; j++) {
+
+      switch (interfaces[i]->readData[j]) {
+      case TEMPERATURE:
+        // Read and set temperature BC
+        if (isQuasi2D3D(interfaces[i]->quasi2D3D)) {
+          consistentScalarRead(interfaces[i]->mappingQuasi2D3D, interfaces[i]->temperatureDataID);
+          setNodeTemperatures(interfaces[i]->mappingQuasi2D3D->bufferScalar3D, interfaces[i]->numNodes, interfaces[i]->xbounIndices, sim->xboun);
+        } else {
+          precicec_readBlockScalarData(interfaces[i]->temperatureDataID, interfaces[i]->numNodes, interfaces[i]->preciceNodeIDs, interfaces[i]->nodeScalarData);
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply adapter/PreciceInterface.c]: Solved by 2740544d83cd23f43ae66ccab829b5ae72ac7dc9 and previous refactoring

File: adapter/PreciceInterface.c [hunk 2]
```diff
@@ -100,32 +96,23 @@ void Precice_Advance(SimulationData *sim)
   printf("Adapter calling advance()...\n");
   fflush(stdout);
 
-  sim->precice_dt = precicec_advance(sim->solver_dt);
+  precicec_advance(sim->solver_dt);
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply adapter/PreciceInterface.c:107]: Clarified in 2740544d83cd23f43ae66ccab829b5ae72ac7dc9

**Review by BenjaminRodenberg — APPROVED** [review]
Time step updates look good now :+1:

I see some possible refactoring for the steady-state case. But would not do this without thorough testing which is a bit out of scope and also not the main goal of this PR.

I'm still not sure if subcycling actually works as expected, but we will see this down the road after some testing. Did you check the (non-subcycling) tutorials we have?

**Comment by IshaanDesai** [comment]
I am testing with the perpendicular-flap tutorial, together with OpenFOAM as the Fluid participant, and CalculiX crashes at the end of the simulation (`time-window: 501, time: 5 of 5`) with the following message:

```bash
 *ERROR: solution seems to diverge; please try 
 automatic incrementation; program stops
 best solution and residuals are in the frd file

corrupted double-linked list
```

The displacement plot looks exactly as before for OpenFOAM-CalculiX:

image

**Comment by fsimonis** [comment]
I cannot reproduce the crash with fluid-nutils and `<max-time value="0.02" />`. Not even when running calculix in a debugger.

**Comment by IshaanDesai** [comment]
Setting `<max-time-windows value="500" />` instead of `max-time value="5"` in the precice-config.xml solves the corruption error.

**Comment by IshaanDesai** [comment]
Merging this as the exit with corruption seems to not be directly related to the changes in this PR.

**Comment by TheLastAirbenderr** [comment]
ccx_preCICE file not generating on compiling v2.20.1

**Pull request merged**